src/middle/Expr.ml

```diff
@@ -0,0 +1,240 @@
+open Core_kernel
+open Common
+open Helpers
+
+(** Pattern and fixed-point of MIR expressions *)
+module Fixed = struct
+  module Pattern = struct
+    type litType = Int | Real | Str [@@deriving sexp, hash, compare]
+
+    type 'a t =
+      | Var of string
+      | Lit of litType * string
+      | FunApp of Fun_kind.t * string * 'a list
+      | TernaryIf of 'a * 'a * 'a
+      | EAnd of 'a * 'a
+      | EOr of 'a * 'a
+      | Indexed of 'a * 'a Index.t list
+    [@@deriving sexp, hash, map, compare, fold]
+
+    let pp pp_e ppf = function
+      | Var varname -> Fmt.string ppf varname
+      | Lit (Str, str) -> Fmt.pf ppf "%S" str
+      | Lit (_, str) -> Fmt.string ppf str
+      | FunApp (StanLib, name, [lhs; rhs])
+        when Option.is_some (Operator.of_string_opt name) ->
+          Fmt.pf ppf "(%a %a %a)" pp_e lhs Operator.pp
+            (Option.value_exn (Operator.of_string_opt name))
+            pp_e rhs
+      | FunApp (_, name, args) ->
+          Fmt.string ppf name ;
+          Fmt.(list pp_e ~sep:Fmt.comma |> parens) ppf args
+      | TernaryIf (pred, texpr, fexpr) ->
+          Fmt.pf ppf {|@[%a@ %a@,%a@,%a@ %a@]|} pp_e pred pp_builtin_syntax "?"
+            pp_e texpr pp_builtin_syntax ":" pp_e fexpr
+      | Indexed (expr, indices) ->
+          Fmt.pf ppf {|@[%a%a@]|} pp_e expr
+            ( if List.is_empty indices then fun _ _ -> ()
+            else Fmt.(list (Index.pp pp_e) ~sep:comma |> brackets) )
+            indices
+      | EAnd (l, r) -> Fmt.pf ppf "%a && %a" pp_e l pp_e r
+      | EOr (l, r) -> Fmt.pf ppf "%a || %a" pp_e l pp_e r
+
+    include Foldable.Make (struct type nonrec 'a t = 'a t
+
+                                  let fold = fold
+    end)
+  end
+
+  include Fixed.Make (Pattern)
+end
+
+(** Expressions without meta data *)
+module NoMeta = struct
+  module Meta = struct
+    type t = unit [@@deriving compare, sexp, hash]
+
+    let empty = ()
+    let pp _ _ = ()
+  end
+
+  include Specialized.Make (Fixed) (Meta)
+
+  let remove_meta x = Fixed.map (Fn.const ()) x
+end
+
+(** Expressions with associated location and type *)
+module Typed = struct
+  module Meta = struct
+    type t =
+      { type_: UnsizedType.t
+      ; loc: Location_span.t sexp_opaque [@compare.ignore]
+      ; adlevel: UnsizedType.autodifftype }
+    [@@deriving compare, create, sexp, hash]
```

Comment:
Why use create instead of record syntax?

Reply:
As above; I would favor hiding the implementation but we expose both for now.